src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+
+                if (resetBlock.Header.HashPrevBlock == this.chain.Genesis.HashBlock)
+                {
+                    resetBlockHash = this.chain.Genesis.HashBlock;
+                    break;
+                }
+
+                resetBlock = await this.blockRepository.GetAsync(resetBlock.Header.HashPrevBlock).ConfigureAwait(false);
+                Guard.NotNull(resetBlock, nameof(resetBlock));
+                resetBlockHash = resetBlock.GetHash();
+            }
+
+            ChainedHeader newTip = this.chain.GetBlock(resetBlockHash);
+            await this.blockRepository.DeleteAsync(newTip.HashBlock, blockStoreResetList).ConfigureAwait(false);
+
+            this.SetStoreTip(newTip);
+
+            this.logger.LogWarning("Block store tip recovered to block '{0}'.", newTip);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Adds a block to the saving queue.
+        /// </summary>
+        /// <param name="blockPair">The block and its chained header pair to be added to pending storage.</param>
+        public void AddToPending(BlockPair blockPair)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(blockPair), blockPair.ChainedHeader);
+
+            this.blocksQueue.Enqueue(blockPair);
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>
+        /// Dequeues the blocks continuously and saves them to the database when max batch size is reached or timer ran out.
+        /// </summary>
+        /// <remarks>Batch is always saved on shutdown.</remarks>
+        private async Task DequeueBlocksContinuouslyAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var batch = new List<BlockPair>();
+
+            Task<BlockPair> dequeueTask = null;
+            Task timerTask = null;
+
+            while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+            {
+                // Start new dequeue task if not started already.
+                dequeueTask = dequeueTask ?? this.blocksQueue.DequeueAsync();
+
+                // Wait for one of the tasks: dequeue or timer (if available) to finish.
+                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip 
```

Comment:
even though it says save batch, it doesn't actually save the batch here, just sets the variable.

Reply:
Yes, it sets the variable. 
The comment like `set this variable to true if **. After that it will be checked and if it's true- it will call saving the batch` sounds weird

Reply:
Yes it does sound weird but also having a comment that says it does something it doesn't is also weird

Reply:
The comment is provided to describe a high level picture, not to describe syntax. 

For example:

```
// true if A is equal to B, false otherwise. 
bool numbersAreEqueal = A == B;
```

this is a bad comment because it basically duplicates what the code says but in other words. 

So if the comment says "save the batch"- the very next line shouldn't actually do exactly that. Here we are making a decision to save the batch, not actually saving it.

Reply:
I don't like that, it is misleading.

Reply:
But it is 4 lines before it does that, this is misleading

Reply:
It is not misleading, it is how comments work – they give you extra information about the context, so that reading the code is easier. This is perfectly fine.

Reply:
I lost it. 
Does what? Set the `saveBatch`? Yes, it may be set to true in several cases. Those cases are described in summary and in inline comments.